src/cs/Ssh.Tcp/Services/StreamForwarder.cs

```diff
@@ -0,0 +1,214 @@
+	protected virtual async Task<bool> ForwardStreamAsync(Stream localStream, Stream remoteStream)
+	{
+		if (localStream == null) throw new ArgumentNullException(nameof(localStream));
+		if (remoteStream == null) throw new ArgumentNullException(nameof(remoteStream));
+
+		using var cancellationSource = new CancellationTokenSource();
+
+		var forwardTask = ForwardStreamAsync(localStream, remoteStream, cancellationSource.Token);
+		var backwardTask = ForwardStreamAsync(remoteStream, localStream, cancellationSource.Token);
+
+		var firstToComplete = await Task.WhenAny(forwardTask, backwardTask)
+			.ConfigureAwait(false);
+		cancellationSource.Cancel();
+
+		await Task.WhenAll(forwardTask, backwardTask).ConfigureAwait(false);
+		return await firstToComplete.ConfigureAwait(false);
+	}
+
+	/// <summary>
+	/// Forwards from a source stream to a destination stream until either the end of the source
+	/// stream is reached, an I/O error occurs, or the cancellation token is cancelled.
+	/// </summary>
+	/// <returns>True if the end of stream was reached (graceful closure) else false</returns>
+	private async Task<bool> ForwardStreamAsync(
+		Stream source,
+		Stream destination,
+		CancellationToken cancellation = default)
+	{
+		byte[] buffer = new byte[8192];
+#if SSH_ENABLE_SPAN
+		var memory = new Memory<byte>(buffer);
+#endif
+		while (!cancellation.IsCancellationRequested)
+		{
+			int count = 0;
+			Exception? readException = null;
+			try
+			{
+#if SSH_ENABLE_SPAN
+				count = await source.ReadAsync(memory, cancellation)
+#else
+				count = await source.ReadAsync(buffer, 0, buffer.Length, cancellation)
+#endif
+					.ConfigureAwait(false);
+			}
+			catch (OperationCanceledException) when (cancellation.IsCancellationRequested)
+			{
+				break;
+			}
+			catch (IOException ex)
+			{
+				readException = ex;
+			}
+			catch (SocketException ex)
+			{
+				readException = ex;
+			}
+			catch (ObjectDisposedException ex)
+			{
+				readException = ex;
+			}
+			catch (SshConnectionException ex)
+			{
+				readException = ex;
+			}
+
+			if (count > 0)
+			{
+				// Do not pass cancellation token to writer because writing may be non-atomic,
+				// and cancelling it may corrupt the stream.
+#if SSH_ENABLE_SPAN
+				await destination.WriteAsync(memory.Slice(0, count), CancellationToken.None)
+#else
+				await destination.WriteAsync(buffer, 0, count, CancellationToken.None)
+#endif
+					.ConfigureAwait(false);
+			}
+			else if (readException == null)
+			{
+				string message = "Channel forwarder reached end of stream.";
+				Trace.TraceEvent(
+					TraceEventType.Verbose, SshTraceEventIds.ChannelClosed, message);
+				break;
+			}
+			else
+			{
+				string message = $"Channel forwarder stream read error: {readException.Message}";
+				Trace.TraceEvent(TraceEventType.Verbose, SshTraceEventIds.ChannelClosed, message);
+				return false;
+			}
+		}
+
+		return true;
```

Comment:
Should we flush the destination stream when we're done forwarding?

Reply:
Done.